File: packages/reliai-python/reliai/instrumentation/langchain.py
from __future__ import annotations

from inspect import iscoroutinefunction
from typing import Any

from ..client import ReliaiClient


def detect_langchain(user_agent: str | None) -> bool:
    return bool(user_agent and "langchain" in user_agent.lower())


def instrument_langchain(client: ReliaiClient) -> bool:
    patched = False
    patched |= _patch_target("langchain.chains.base", "Chain", "invoke", client, "langchain.chain.invoke", {"framework": "langchain", "component": "chain"})
    patched |= _patch_target("langchain.llms.base", "LLM", "invoke", client, "langchain.llm.invoke", {"framework": "langchain", "component": "llm"})
    patched |= _patch_target("langchain_core.language_models.llms", "LLM", "invoke", client, "langchain.llm.invoke", {"framework": "langchain", "component": "llm"})
    patched |= _patch_target("langchain.tools.base", "Tool", "run", client, "langchain.tool.run", {"framework": "langchain", "component": "tool"})
    patched |= _patch_target("langchain_core.tools", "Tool", "run", client, "langchain.tool.run", {"framework": "langchain", "component": "tool"})
    return patched


def _patch_target(
    module_name: str,
    class_name: str,
    method_name: str,
    client: ReliaiClient,
    span_name: str,
    metadata: dict[str, Any],
) -> bool:
    try:
        module = __import__(module_name, fromlist=[class_name])
    except Exception:
        return False
    candidate = getattr(module, class_name, None)
    if not isinstance(candidate, type):
        return False
    return _patch_method(candidate, method_name, client, span_name, metadata)


def _patch_method(cls: type[Any], method_name: str, client: ReliaiClient, span_name: str, metadata: dict[str, Any]) -> bool:
    original = getattr(cls, method_name, None)
    if not callable(original) or getattr(original, "__reliai_patched__", False):
        return False

    if iscoroutinefunction(original):

        async def wrapped(self: Any, *args: Any, **kwargs: Any):
            extra = _extra_metadata(self, metadata)
            with client.span(span_name, extra):
                return await original(self, *args, **kwargs)

    else:

        def wrapped(self: Any, *args: Any, **kwargs: Any):
            extra = _extra_metadata(self, metadata)
            with client.span(span_name, extra):
                return original(self, *args, **kwargs)

    setattr(wrapped, "__reliai_patched__", True)
    setattr(cls, method_name, wrapped)
    return True


def _extra_metadata(instance: Any, metadata: dict[str, Any]) -> dict[str, Any]:
    extra = dict(metadata)
    tool_name = getattr(instance, "name", None)
    if isinstance(tool_name, str) and tool_name:
        extra["tool_name"] = tool_name
    return extra

File: packages/reliai-node/src/instrumentation/langchain.ts
import type { ReliaiClient } from "../client.ts";

function getHeader(headers: Headers | Record<string, string> | string[][], key: string): string | null {
  if (headers instanceof Headers) {
    return headers.get(key);
  }
  if (Array.isArray(headers)) {
    const match = headers.find(([candidate]) => candidate.toLowerCase() === key.toLowerCase());
    return match?.[1] ?? null;
  }
  const direct = headers[key] ?? headers[key.toLowerCase()];
  return typeof direct === "string" ? direct : null;
}

export function detectFrameworkFromHeaders(headers: Headers | Record<string, string> | string[][]): string | null {
  const userAgent = getHeader(headers, "user-agent") ?? getHeader(headers, "x-stainless-user-agent") ?? "";
  const normalized = userAgent.toLowerCase();
  if (normalized.includes("langchain")) {
    return "langchain";
  }
  if (normalized.includes("llamaindex")) {
    return "llamaindex";
  }
  return null;
}

function isFunction(value: unknown): value is (...args: unknown[]) => unknown {
  return typeof value === "function";
}

function patchMethod(
  target: Record<string, unknown>,
  methodName: string,
  wrap: (original: (...args: unknown[]) => unknown) => (...args: unknown[]) => unknown,
): void {
  const original = target[methodName];
  if (!isFunction(original) || (original as { __reliai_patched__?: boolean }).__reliai_patched__) {
    return;
  }
  const patched = wrap(original);
  (patched as { __reliai_patched__?: boolean }).__reliai_patched__ = true;
  target[methodName] = patched;
}

async function patchLangChainModule(
  client: ReliaiClient,
  moduleName: string,
): Promise<void> {
  try {
    const imported = await import(moduleName);
    const candidates = Object.values(imported).filter((value): value is Record<string, unknown> => Boolean(value) && typeof value === "function");
    for (const candidate of candidates) {
      const prototype = candidate.prototype as Record<string, unknown> | undefined;
      if (!prototype) {
        continue;
      }
      patchMethod(prototype, "run", (original) => function patchedRun(this: unknown, ...args: unknown[]) {
        return client.span("llm_call", () => original.apply(this, args), {
          span_type: "llm",
          framework: "langchain",
        });
      });
      patchMethod(prototype, "getRelevantDocuments", (original) => function patchedRetriever(this: unknown, ...args: unknown[]) {
        return client.span("retrieval", () => original.apply(this, args), {
          span_type: "retrieval",
          framework: "langchain",
        });
      });
      patchMethod(prototype, "invoke", (original) => function patchedInvoke(this: unknown, ...args: unknown[]) {
        const toolName =
          this && typeof this === "object" && "name" in (this as Record<string, unknown>) && typeof (this as Record<string, unknown>).name === "string"
            ? ((this as Record<string, unknown>).name as string)
            : undefined;
        return client.span("tool_call", () => original.apply(this, args), {
          span_type: "tool",
          framework: "langchain",
          ...(toolName ? { tool_name: toolName } : {}),
        });
      });
    }
  } catch {
    return;
  }
}

export async function instrumentLangChain(client: ReliaiClient): Promise<void> {
  await Promise.allSettled([
    patchLangChainModule(client, "langchain/chains"),
    patchLangChainModule(client, "@langchain/core/retrievers"),
    patchLangChainModule(client, "@langchain/core/tools"),
  ]);
}
